tests/multiple_database/tests.py

```diff
@@ -2082,3 +2082,120 @@ def test_reverse_m2m_update(self):
         self.assertEqual(e.mode, RouterUsed.WRITE)
         self.assertEqual(e.model, Book)
         self.assertEqual(e.hints, {'instance': auth})
+
+
+class NoRelationRouter:
+    """
+    Router for testing that relationships between foreign objects are enforced
+    """
+    def allow_relation(self, obj1, obj2, **hints):
+        # Cause all foreign relations to fail
+        return False
+
+
+@override_settings(DATABASE_ROUTERS=[NoRelationRouter()])
+class AllowRelationTests(TestCase):
```

Comment:
This has been reduced to be only test cases that fail prior to this fix.

Reply:
Still, I wonder if so many tests are needed. I think it would be enough to have one test for ForwardManyToOneDescriptor and ReverseOneToOneDescriptor. In other words, every place that the descriptor is called doesn't need to be tested. Testing the each descriptor once is enough, I think.

Please use the context manager version of assertRaisesMessage and follow the style of similar assertions in this file.

Reply:
The multitude of constructor-variant tests could be removed, but the messed up logic that used to surround `allow_relation()` results in different behavior based on the state of the instance or value objects (whether they are associated with a database or not at the time of the check).

Ensuring those test cases are covered is needed, but that also involves multiple tests per descriptor type.

Reply:
If the testing strategy is to ensure the new implementation is properly tested (not necessarily the all conditions of the old implementation), then I agree that one test per descriptor is sufficient.

I've updated the tests further. Now there are only two.